src/lib.rs

```diff
@@ -30,45 +31,62 @@ pub fn load_happs_yaml(path: impl AsRef<Path>) -> Result<Vec<Happ>> {
 }
 
 pub async fn install_happs(happ_list: &[Happ], config: &Config) -> Result<()> {
-    let admin_websocket = AdminWebsocket::connect(config.admin_port)
+    let mut admin_websocket = AdminWebsocket::connect(config.admin_port)
         .await
         .context("failed to connect to holochain")?;
+
+    if let Err(error) = admin_websocket.attach_app_interface(config.happ_port).await {
+        warn!(port = ?config.happ_port, ?error, "failed to start interface, maybe it's already up?");
+    }
+
+    let agent_key = admin_websocket
+        .generate_agent_pubkey()
+        .await
+        .context("failed to generate agent key")?;
+
+    let installed_happs = Arc::new(
+        admin_websocket
+            .list_installed_happs()
+            .await
+            .context("failed to get installed hApps")?,
+    );
+
     let futures: Vec<_> = happ_list
         .iter()
         .map(|happ| {
             let mut admin_websocket = admin_websocket.clone();
+            let agent_key = agent_key.clone();
+            let installed_happs = Arc::clone(&installed_happs);
             async move {
-                let mut agent_websocket = admin_websocket.clone();
-                let install_happ = agent_websocket
-                    .generate_agent_pubkey()
-                    .and_then(|agent_key| {
-                        admin_websocket.install_happ(happ, agent_key, config.happ_port)
-                    });
                 let install_ui = install_ui(happ, config);
-                futures::try_join!(install_happ, install_ui)
+                if installed_happs.contains(&happ.id_with_version()) {
+                    info!(?happ.installed_app_id, "already installed, just downloading UI");
+                    install_ui.await
+                } else {
+                    let install_happ = admin_websocket.install_happ(happ, agent_key);
+                    futures::try_join!(install_happ, install_ui).map(|_| ())
+                }
             }
         })
         .collect();
-    let _: Vec<_> = try_join_all(futures)
-        .await
-        .context("failed to install some hApps")?;
+    let _: Vec<_> = future::join_all(futures).await;
 
-    info!("all hApps installed");
+    info!("finished installing hApps");
     Ok(())
 }
 
 #[instrument(
     err,
     skip(happ, config),
-    fields(?happ.app_id)
+    fields(?happ.installed_app_id)
 )]
 async fn install_ui(happ: &Happ, config: &Config) -> Result<()> {
     let mut ui_archive = download_file(&happ.ui_url)
         .await
         .context("failed to download UI archive")?;
-    let unpack_path = config.ui_store_folder.join(&happ.app_id);
+    let unpack_path = config.ui_store_folder.join(&happ.installed_app_id);
     extract_zip(ui_archive.as_file_mut(), unpack_path).context("failed to extract UI archive")?;
```

Comment:
I noted that when I saw this error, I didn't get to see the underlying system error.  Is there something I have to do to get to see it?